pkg/consuldp/lifecycle_test.go

```diff
@@ -0,0 +1,195 @@
+		},
+		"connection draining enabled with grace period": {
+			// This should immediately send "Connection: close" to inbound HTTP1
+			// connections, GOAWAY to inbound HTTP2, and terminate connections on
+			// request completion.
+			// Outbound connections should be allowed until the grace period has
+			// elapsed, then any remaining open connections should be closed and new
+			// outbound connections should start being rejected until pod termination.
+			shutdownDrainListeners: true,
+			shutdownGracePeriod:    5,
+		},
+		"custom graceful shutdown path and port": {
+			shutdownDrainListeners: true,
+			shutdownGracePeriod:    5,
+			gracefulShutdownPath:   "/quit-nicely",
+			// TODO: should this be random or use freeport? logic disallows passing
+			// zero value explicitly
+			gracefulPort: 23108,
+		},
+	}
+	for name, c := range cases {
+		c := c
+		log.Printf("config = %v", c)
+
+		t.Run(name, func(t *testing.T) {
+			cfg := Config{
+				Envoy: &EnvoyConfig{
+					AdminBindAddress:       envoyAdminAddr,
+					AdminBindPort:          envoyAdminPort,
+					ShutdownDrainListeners: c.shutdownDrainListeners,
+					ShutdownGracePeriod:    c.shutdownGracePeriod,
+					GracefulShutdownPath:   c.gracefulShutdownPath,
+					GracefulPort:           c.gracefulPort,
+				},
+			}
+			m := NewLifecycleConfig(&cfg, &mockProxy{})
+
+			require.NotNil(t, m)
+			require.NotNil(t, m.proxy)
+			require.NotNil(t, m.errorExitCh)
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			err := m.startLifecycleManager(ctx)
+			require.NoError(t, err)
+
+			// Have consul-dataplane's lifecycle server start on an open port
+			// and figure out what port was used so we can make requests to it.
+			// Conveniently, this seems to wait until the server is ready for requests.
+			portCh := make(chan int, 1)
+			if c.gracefulPort == 0 {
+				m.lifecycleServer.Addr = "127.0.0.1:0"
+			}
+			m.lifecycleServer.BaseContext = func(l net.Listener) context.Context {
+				portCh <- l.Addr().(*net.TCPAddr).Port
+				return context.Background()
+			}
+
+			var port int
+			select {
+			case port = <-portCh:
+			case <-time.After(5 * time.Second):
+			}
+
+			// Check lifecycle server graceful port configuration
+			if c.gracefulPort != 0 {
+				require.Equal(t, c.gracefulPort, port, "failed to set lifecycle server port")
+			} else {
+				require.NotEqual(t, 0, port, "failed to figure out lifecycle server port")
+			}
+			log.Printf("port = %v\n", port)
+
+			// Check lifecycle server graceful shutdown path configuration
+			if c.gracefulShutdownPath != "" {
+				require.Equal(t, m.gracefulShutdownPath, c.gracefulShutdownPath, "failed to set lifecycle server graceful shutdown HTTP endpoint path")
+			}
+
+			// Check lifecycle server graceful shutdown path configuration
+			url := fmt.Sprintf("http://127.0.0.1:%d%s", port, m.gracefulShutdownPath)
+			log.Printf("sending request to %s\n", url)
+
+			resp, err := http.Get(url)
```

Comment:
Could possibly add some way to check that the Envoy process stays alive for `c.shutdownGracePeriod` seconds here.